base_layer/core/src/proof_of_work/median_timestamp.rs

```diff
@@ -20,13 +20,32 @@
 // WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 // USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-use crate::{chain_storage::ChainStorageError, proof_of_work::error::DifficultyAdjustmentError};
-use derive_error::Error;
+use crate::blocks::blockheader::BlockHeader;
+use log::*;
+use tari_crypto::tari_utilities::epoch_time::EpochTime;
 
-#[derive(Debug, Error, Clone, PartialEq)]
-pub enum DiffAdjManagerError {
-    DifficultyAdjustmentError(DifficultyAdjustmentError),
-    ChainStorageError(ChainStorageError),
-    EmptyBlockchain,
-    PoisonedAccess,
+pub const LOG_TARGET: &str = "c::pow::median_timestamp";
+
+/// Returns the median timestamp for the provided header set.
+pub fn get_median_timestamp(headers: Vec<BlockHeader>) -> Option<EpochTime> {
+    if headers.len() == 0 {
+        return None;
+    }
+    let height = headers.last().expect("Header set should not be empty").height;
+    debug!(target: LOG_TARGET, "Calculating median timestamp to height:{}", height);
+    let mut timestamps = headers.iter().map(|h| h.timestamp).collect::<Vec<_>>();
+    timestamps.sort();
+    trace!(target: LOG_TARGET, "Sorted median timestamps: {:?}", timestamps);
+    // Calculate median timestamp
+    let mid_index = timestamps.len() / 2;
+    // let median_timestamp=if timestamps.len()%2==0 {
+    // (timestamps[mid_index-1]+timestamps[mid_index])/2
+    // }
+    // else { timestamps[mid_index] };
+    let median_timestamp = timestamps[mid_index];
```

Comment:
Is it time to switch the median calculation to a true median calculation as specified in #1118? This change will be consensus breaking.